[X86] Align f128 and i128 to 16 bytes when passing on x86-32

tgross35 opened the pull request:
The i386 psABI specifies that `__float128` has 16 byte alignment and
must be passed on the stack; however, LLVM currently stores it in a
stack slot that has an offset of 4. Add a custom lowering to correct
this alignment to 16-byte.

i386 does not specify an `__int128`, but it seems reasonable to keep the
same behavior as `__float128` so this is changed as well. There also
isn't a good way to distinguish whether a set of four registers came
from an integer or a float.

The main test demonstrating this change is `store_perturbed` in
`llvm/test/CodeGen/X86/i128-fp128-abi.ll`.

Referenced ABI: https://gitlab.com/x86-psABIs/i386-ABI/-/wikis/uploads/14c05f1b1e156e0e46b61bfa7c1df1e2/intel386-psABI-2020-08-07.pdf
Fixes: https://github.com/llvm/llvm-project/issues/77401

arsenm commented:
> I'm not sure why this seems to have no effect; the tests at llvm/test/CodeGen/X86/fp128-abi.ll should fail since they need to be updated, but there is no change in codegen. I wonder if this is getting fixed up somewhere.

fp128 is not a legal type, so this won't reach here (this is a super frustrating detail of how calling convention handling is implemented, it would be easier if the backend directly consumed the original type)

tgross35 commented:
Should it be made a legal type? I’m not sure what the exact definition of that is, but it is in the i386 ABI as of recent versions to be returned and (presumably, since I haven’t seen otherwise) passed in memory.

arsenm commented:
> Should it be made a legal type? 
No. There are no f128 registers or operations

tgross35 commented:
How does stack alignment get handled for non-legal types, or where does this need to get fixed up? I’m surprised passing doesn’t use the type’s natural alignment.

arsenm commented:
> How does stack alignment get handled for non-legal types, or where does this need to get fixed up? I’m surprised passing doesn’t use the type’s natural alignment.

The type is broken up into legal pieces according to getRegisterTypeForCallingConv + getNumRegistersForCallingConv, and those pieces are processed by the call lowering. This code is entirely disconnected from the decision of where to pass the value, in a register or at some stack offset

tgross35 committed:
[X86] Align f128 and i128 to 16 bytes when passing on x86-32

The i386 psABI specifies that `__float128` has 16 byte alignment and
must be passed on the stack; however, LLVM currently stores it in a
stack slot that has an offset of 4. Add a custom lowering to correct
this alignment to 16-byte.

i386 does not specify an `__int128`, but it seems reasonable to keep the
same behavior as `__float128` so this is changed as well.

Fixes: https://github.com/llvm/llvm-project/issues/77401

tgross35 commented on llvm/lib/Target/X86/X86ISelLoweringCall.cpp:
```diff
@@ -237,9 +237,18 @@ EVT X86TargetLowering::getSetCCResultType(const DataLayout &DL,
 bool X86TargetLowering::functionArgumentNeedsConsecutiveRegisters(
     Type *Ty, CallingConv::ID CallConv, bool isVarArg,
     const DataLayout &DL) const {
-  // i128 split into i64 needs to be allocated to two consecutive registers,
-  // or spilled to the stack as a whole.
-  return Ty->isIntegerTy(128);
+  // On x86-64 i128 is split into two i64s and needs to be allocated to two
+  // consecutive registers, or spilled to the stack as a whole. On x86-32 i128
+  // is split to four i32s and never actually passed in registers, but we use
+  // the consecutive register mark to match it in TableGen.
+  if (Ty->isIntegerTy(128))
+    return true;
+
+  // On x86-32, fp128 acts the same as i128.
+  if (Subtarget.is32Bit() && Ty->isFP128Ty())
+    return true;
+
+  return false;
```
I'm not sure if this is the right approach; `functionArgumentNeedsConsecutiveRegisters` doesn't really seem like the correct thing because the type will never be passed in regesters, but I can't come up with another way to "mark" a register set to indicate that it needs custom lowering. Is there a better way to do this?

Cc @nikic since I think you rewrote the x86-64 custom lowering a couple of times.

tgross35 replied:
This should probably also match vector types somehow because `_m64`, `__m128`, `__m256`, and `__m512` are specified to have an alignment of 8, 16, 32, and 64 bytes, respectively.

nikic replied:
Yeah, unfortunately the CC lowering is pretty limited in the information it can access, and backends have a lot of different hacks to work around that. I'd like to improve that, but this looks like an acceptable hack for now.

tgross35 commented on llvm/test/CodeGen/X86/all-ones-vector.ll:
```diff
@@ -272,7 +272,7 @@ define <64 x i8> @allones_v64i8() nounwind {
 ; AVX512-LABEL: allones_v64i8:
 ; AVX512:       # %bb.0:
-; AVX512-NEXT:    vpternlogd $255, %zmm0, %zmm0, %zmm0
+; AVX512-NEXT:    vpternlogd {{.*#+}} zmm0 = -1
```
I am not sure what caused the change here

phoebewang replied:
This is due to comment change for vpternlogd. Not related to the change here.

tgross35 replied:
Thanks, looks like I was able to drop this.

tgross35 commented:
I also brought up discussion about specifying an ABI for __int128 https://groups.google.com/u/1/g/ia32-abi/c/TuMAt7mwbIU

tgross35 commented:
Rebased to pick up a relevant test change. Also put up a separate PR with the pretest so that can get its own commit https://github.com/llvm/llvm-project/pull/148753.

nikic reviewed (approved):
LGTM from my side.

Looks like this needs a rebase.

tgross35 committed:
Remove unneeded auto test updates

tgross35 committed:
Remove accidental duplicate comment

tgross35 commented on llvm/lib/Target/X86/X86CallingConv.cpp:
```diff
@@ -374,5 +374,37 @@ static bool CC_X86_64_I128(unsigned &ValNo, MVT &ValVT, MVT &LocVT,
+  if (!ArgFlags.isInConsecutiveRegsLast())
+    return true;
+
+  unsigned NumRegs = PendingMembers.size();
```
It looks like somebody got this already https://github.com/llvm/llvm-project/commit/fe1941967267e472

tgross35 commented on llvm/lib/Target/X86/X86CallingConv.cpp:
```diff
@@ -374,5 +374,37 @@ static bool CC_X86_64_I128(unsigned &ValNo, MVT &ValVT, MVT &LocVT,
+    return true;
+
+  unsigned NumRegs = PendingMembers.size();
+  assert(NumRegs == 4 && "Should have two parts");
```
Fix in https://github.com/llvm/llvm-project/pull/149386

nikic commented:
/cherry-pick a78a0f8d204393a0cce367b63395bad90311c1b8 fe1941967267e472f7eee15b43712bdfa2b63544

tgross35 commented:
Created https://github.com/llvm/llvm-project/pull/150746
